### PR TITLE
Fix issues with RLPExceptions

### PR DESCRIPTION
**Pull request by Gabriel-Trintinalia**
## PR description

This PR fixes two issues with RLP encoding:

### 1- BlockBody

For the BlockBody RLP serialisation, when an empty body `[]` is sent, a breach of protocol is triggered and Besu disconnects the peer. That causes Besu to disconnect peers with a good reputation. This PR adds an option in the readFrom method to allow an empty BlockBody whenever a `[]` is found. The message will be verified later and marked as invalid anyway, but besu will not disconnect the peer.

### 2 -NewPooledTransactionHashesMessage

`NewPooledTransactionHashesMessage` on Eth/68 has an issue with the array of types (`byte`):

`[[type_0: B_1, type_1: B_1, ...], [size_0: B_4, size_1: B_4, ...], [hash_0: B_32, hash_1: B_32, ...]]`

The current implementation serializes it as:
```
[
      ["0x01","0x02"]
      ["0x00000001","0x00000002"],
      ["0x0000000000000000000000000000000000000000000000000000000000000001",
       "0x0000000000000000000000000000000000000000000000000000000000000002"]
]
```
The other clients send the first array of this message as one element instead of an array:
```
[
      ["0x0102"]
      ["0x00000002","0x00000003"],
      ["0x0000000000000000000000000000000000000000000000000000000000000002",
       "0x0000000000000000000000000000000000000000000000000000000000000003"]
]
```
This PR fixes the encoding to match the other client messages.

 see #5056

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Acceptance Tests (Non Mainnet)

- [ ] I have considered running `./gradlew acceptanceTestNonMainnet` locally if my PR affects non-mainnet modules.

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/BlockBody.java [hunk 1]
```diff
@@ -97,7 +97,19 @@ public void writeTo(final RLPOutput output) {
...
+  public static BlockBody readFrom(
+      final RLPInput input,
+      final BlockHeaderFunctions blockHeaderFunctions,
+      final boolean allowEmptyBody) {
     input.enterList();
+    if (input.isEndOfCurrentList() && allowEmptyBody) {
+      // empty block [] -> Return empty body.
+      input.leaveList();
+      return empty();
+    }
```
Comment (Gabriel-Trintinalia) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/BlockBody.java:112]: Handle empty body

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/encoding/TransactionAnnouncementDecoder.java [hunk 2]
```diff
@@ -69,13 +70,14 @@ private static List<TransactionAnnouncement> decodeForEth66(final RLPInput input
...
-              final int type = rlp.readByte() & 0xff;
-              return type == 0 ? TransactionType.FRONTIER : TransactionType.of(type);
-            });
-    final List<Long> sizes = input.readList(RLPInput::readUnsignedInt);
+
+    final List<TransactionType> types = new ArrayList<>();
+    final byte[] bytes = input.readBytes().toArray();
+    for (final byte b : bytes) {
+      types.add(b == 0 ? TransactionType.FRONTIER : TransactionType.of(b));
+    }
```
Comment (Gabriel-Trintinalia) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/encoding/TransactionAnnouncementDecoder.java:78]: Fix decoding

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/encoding/TransactionAnnouncementEncoder.java [hunk 3]
```diff
@@ -76,30 +76,42 @@ private static Bytes encodeForEth66(final List<Transaction> transactions) {
...
     // Check if lists have the same size
-    if (!(types.size() == hashes.size() && hashes.size() == sizes.size())) {
+    if (!(types.length == hashes.size() && hashes.size() == sizes.size())) {
       throw new IllegalArgumentException(
           "Hashes, sizes and types must have the same number of elements");
     }
     out.startList();
-    out.writeList(
-        types, (h, w) -> w.writeByte(h == TransactionType.FRONTIER ? 0x00 : h.getSerializedType()));
+    out.writeBytes(Bytes.wrap((types)));
```
Comment (Gabriel-Trintinalia) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/encoding/TransactionAnnouncementEncoder.java:114]: Fix encoding

**Review by pinges — APPROVED** [review]
LGTM

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/NewPooledTransactionHashesMessageProcessorTest.java [hunk 4]
```diff
@@ -289,7 +289,7 @@ public void shouldEncodeTransactionsCorrectly_Eth68() {
   public void shouldDecodeBytesCorrectly_Eth68() {
     /*
      * [
-     * ["0x00","0x01","0x02"]
+     * "0x00102"]
```
Comment (pinges) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/NewPooledTransactionHashesMessageProcessorTest.java:292]: Is there a 0 missing at the start?

Comment (Gabriel-Trintinalia) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/NewPooledTransactionHashesMessageProcessorTest.java]: good eye! fixed

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/messages/BlockBodiesMessageTest.java [hunk 5]
```diff
@@ -65,16 +81,37 @@ public void blockBodiesRoundTrip() throws IOException {
...
+    final Bytes bytes = Bytes.fromHexString("0xc0");
+    final BlockBody empty = BlockBody.readFrom(RLP.input(bytes), null, true);
+    Assertions.assertThat(empty.isEmpty()).isTrue();
+  }
+
+  @Test
+  public void shouldThrowRLPExceptionIfNotAllowedEmptyBody() {
+    final Bytes bytes = Bytes.fromHexString("0xc0");
+    final BlockHeaderFunctions blockHeaderFunctions =
+        ScheduleBasedBlockHeaderFunctions.create(protocolSchedule);
```
Comment (macfarla) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/messages/BlockBodiesMessageTest.java:110]: is the blockHeaderFunctions used?

Comment (macfarla) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/messages/BlockBodiesMessageTest.java:110]: it's not created for the other test shouldNotThrowRLPExceptionIfAllowedEmptyBody

**Review by macfarla — APPROVED** [review]
LGTm.
is there a related hive test for the eth/68 encoding issue?

**Pull request merged**